File: bin/format.rb
#!/usr/bin/env ruby

HELP=<<EOS
format markdown line break style

$ format.rb target.md
EOS

def label(line)
  case line
  when "";                      [:br,     "\n"]
  when /^---/;                  [:meta,   line] # zenn
  when /^\#{1,6} .*/;           [:header, line]
  when /^ *- .*/;               [:ul,     line]
  when /^ *\* .*/;              [:ul,     line]
  when /^ *\+ .*/;              [:ol,     line]
  when /^: .*/;                 [:dl,     line]
  when /^ *\d+\. .*/;           [:num,    line]
  when /^> /;                   [:blockquote, line]
  when /```.*/;                 [:code,   line]
  when /^\|.*/;                 [:table,  line]
  when /^<(?!http(s*):\/\/).*/; [:html,   line]
  when /^:::.*/;                [:message,line] # zenn
  else                          [:p,      line]
  end
end

def tokenize(data)
  data.split("\n").map(&method(:label))
end


class State
  attr_reader :acc
  def initialize
    @state = :start
    @acc = []
  end

  def start((key, val))
    if key == :meta
      # --- で始まるならメタ情報
      @acc << val
      @acc << "\n"
      @state = :meta
    else
      # 始まらないなら title があるはず
      @state = :title
      title([key, val])
    end
  end

  def meta((key, val))
    if key == :meta
      # meta が終わる
      @acc << val
      @acc << "\n\n"
      @state = :title
    else
      # そのまま
      @acc << val
      @acc << "\n"
    end
  end

  def title((key, val))
    if key == :header
      @acc << val
      @state = :body
    elsif key == :br
      # title より上の改行は無視
    else
      pp key, val
      raise "state error"
    end
  end

  def body((key, val))
    case key
    when :header;
      @acc << "\n\n\n"
      @acc << val
    when :p;
      @acc << "\n\n"
      @acc << val
    when :ul;
      @acc << "\n\n"
      @acc << val
      @state = :ul
    when :ol;
      @acc << "\n\n"
      @acc << val
      @state = :ol
    when :dl;
      @acc << "\n"
      @acc << val.sub(/: +/, ": ")
      @state = :dl
    when :num;
      @acc << "\n\n"
      @acc << val
      @state = :num
    when :blockquote;
      @acc << "\n\n"
      @acc << val
      @state = :blockquote
    when :code;
      @acc << "\n\n\n"
      @acc << val
      @state = :code
    when :table;
      @acc << "\n\n"
      @acc << val
      @state = :table
    when :html;
      @acc << "\n\n\n"
      @acc << val
      @acc << "\n"
      @state = :body
    when :message;
      @acc << "\n\n"
      @acc << val
      @state = :message
    when :br;
      # do nothing
    else
      pp key, val
      raise "state error"
    end
  end

  def ul((key, val))
    if key == :ul
      @acc << "\n"
      @acc << val
    else
      @state = :body
    end
  end

  def ol((key, val))
    if key == :ol
      @acc << "\n"
      @acc << val
    else
      @state = :body
    end
  end

  def dl((key, val))
    # dl は : で始まる行で終わるので
    # ここでやることはない
    if key == :dl
    else
      @state = :body
    end
  end

  def num((key, val))
    if key == :num
      @acc << "\n"
      @acc << val
    else
      @state = :body
      process([key, val])
    end
  end

  def code((key, val))
    if key == :code
      @acc << "\n"
      @acc << val
      @state = :body
    elsif key == :br
      @acc << val
    else
      @acc << "\n"
      @acc << val
    end
  end

  def blockquote((key, val))
    if key == :blockquote
      @acc << "\n"
      @acc << val
    elsif key == :br
      @state = :body
    else
      @acc << "\n"
      @acc << val
      @state = :body
    end
  end

  def table((key, val))
    if key == :table
      @acc << "\n"
      @acc << val
    else
      @acc << "\n"
      @state = :body
    end
  end

  def message((key, val))
    if key == :message
      @acc << "\n"
      @acc << val
      @state = :body
    elsif key == :br
      @acc << val
    else
      @acc << "\n"
      @acc << val
    end
  end

  def process(tuple)
    self.send(@state, tuple)
  end
end


def main(target)
  path = File.absolute_path(target)
  data = File.read(path)

  data.gsub!(/ +$/, '') # remove trailing space

  state = State.new()
  tokenize(data).each{|tuple|
    # p tuple
    state.process(tuple)
  }

  result = state.acc.join("") << "\n"

  # puts result
  File.write(path, result)
end


if ARGV.empty?
  puts HELP
  exit 0
end

Dir.glob(ARGV).each{|target|
  main(target)
}

File: bin/format.erl
#!/usr/bin/env escript
-module(format).

-mode(compile).
-compile(export_all).
-behaviour(gen_statem).
-include("logger.hrl").

init(Arg) ->
    {ok, title, Arg}.

callback_mode() ->
    state_functions.

terminate(_Reason, _State, _Data) ->
    void.

code_change(_Vsn, State, Data, _Extra) ->
    {ok, State, Data}.


start() ->
    gen_statem:start({local, ?MODULE}, ?MODULE, "", []).

stop() ->
    gen_statem:stop(?MODULE).

call(Data) ->
    gen_statem:call(?MODULE, Data).

title({call, From}, {header, Line}, _State) ->
    State = _State ++ Line,
    {next_state, intro, State, [{reply, From, State}]}.

intro({call, From}, {br, Line}, _State) ->
    State = _State ++ Line,
    {next_state, intro, State, [{reply, From, State}]};

intro({call, From}, {header, Line}, _State) ->
    State = _State ++ Line,
    {next_state, body, State, [{reply, From, State}]}.

body({call, From}, {header, Line}, _State) ->
    State = _State ++ "\n\n" ++ Line,
    {keep_state, State, [{reply, From, State}]};

body({call, From}, {p, Line}, _State) ->
    State = _State ++ "\n" ++ Line,
    {keep_state, State, [{reply, From, State}]};

body({call, From}, {ul, Line}, _State) ->
    State = _State ++ "\n" ++ Line,
    {next_state, ul, State, [{reply, From, State}]};

body({call, From}, {ol, Line}, _State) ->
    State = _State ++ "\n" ++ Line,
    {next_state, ol, State, [{reply, From, State}]};

body({call, From}, {num, Line}, _State) ->
    State = _State ++ "\n" ++ Line,
    {next_state, num, State, [{reply, From, State}]};

body({call, From}, {code, Line}, _State) ->
    State = _State ++ "\n\n" ++ Line,
    {next_state, code, State, [{reply, From, State}]};

body({call, From}, {table, Line}, _State) ->
    State = _State ++ "\n\n" ++ Line,
    {next_state, table, State, [{reply, From, State}]};

body({call, From}, {br, _}, State) ->
    {keep_state, State, [{reply, From, State}]}.




ul({call, From}, {ul, Line}, _State) ->
    State = _State ++ Line,
    {keep_state, State, [{reply, From, State}]};

ul({call, From}, _, State) ->
    {next_state, body, State, [{reply, From, State}, postpone]}.


ol({call, From}, {ol, Line}, _State) ->
    State = _State ++ Line,
    {keep_state, State, [{reply, From, State}, postpone]};

ol({call, From}, _, State) ->
    {next_state, body, State, [{reply, From, State}]}.


num({call, From}, {num, Line}, _State) ->
    State = _State ++ Line,
    {keep_state, State, [{reply, From, State}]};

num({call, From}, _, State) ->
    {next_state, body, State, [{reply, From, State}, postpone]}.


code({call, From}, {code, Line}, _State) ->
    State = _State ++ Line,
    {next_state, body, State, [{reply, From, State}]};

code({call, From}, {_, Line}, _State) ->
    State = _State ++ Line,
    {keep_state, State, [{reply, From, State}]}.

table({call, From}, {table, Line}, _State) ->
    State = _State ++ Line,
    {keep_state, State, [{reply, From, State}]};

table({call, From}, {_, Line}, _State) ->
    State = _State ++ Line,
    {next_state, body, State, [{reply, From, State}]}.



%% labeling token
label(Line) ->
    label(Line, br).

label(Line, br) ->
    case Line =:= "\n" of
        true -> {br, Line};
        false -> label(Line, header)
    end;

label(Line, header) ->
    case re:run(Line, "^#{1,6} .*") of
        {match, _} -> {header, Line};
        nomatch -> label(Line, ul)
    end;

label(Line, ul) ->
    case re:run(Line, "^ *- .*") of
        {match, _} -> {ul, Line};
        nomatch -> label(Line, ol)
    end;

label(Line, ol) ->
    case re:run(Line, "^ *\\+ .*") of
        {match, _} -> {ol, Line};
        nomatch -> label(Line, num)
    end;

label(Line, num) ->
    case re:run(Line, "^ *\\d+\\. .*") of
        {match, _} -> {num, Line};
        nomatch -> label(Line, code)
    end;

label(Line, code) ->
    case re:run(Line, "```.*") of
        {match, _} -> {code, Line};
        nomatch -> label(Line, table)
    end;

label(Line, table) ->
    case re:run(Line, "^\\|") of
        {match, _} ->
            {table, Line};
        nomatch -> label(Line, p)
    end;

label(Line, p) ->
    {p, Line}.


%% tokenize
token(<<"\n", Rest/binary>>, Curr, Acc) ->

    Line = binary_to_list(<<Curr/binary, "\n">>),

    Labeled = label(Line),

    token(Rest, <<>>, [Labeled | Acc]);

token(<<C, Rest/binary>>, Curr, Acc) ->
    token(Rest, <<Curr/binary, C>>, Acc);

token(<<>>, <<>>, Acc) ->
    lists:reverse(Acc).

token(Bin) ->
    token(Bin, <<>>, []).


main([File]) ->
    {ok, Bin} = file:read_file(File),

    Lines = (token(Bin)),

    start(),

    lists:foreach(fun(Line) ->
                          call(Line)
                  end, Lines),


    Result = call({br, "\n"}),

    file:write_file(File, Result),
    io:format("~s", [Result]),

    ok.
